resource-thief.c

```diff
@@ -0,0 +1,361 @@
+	if (mode == CONSUME_FDS) {
+		resource = "linux dmabuf fds";
+	} else if (mode == CONSUME_MEMORY) {
+		resource = "wl_region points";
+	} else if (mode == CONSUME_MAPS) {
+		resource = "shm pool mappings";
+	}
+
+	fprintf(stderr, "Trying to create %"PRIu32" %s\n", resource_count, resource);
+
+	struct connection *conn = calloc(1, sizeof(struct connection));
+	if (!conn) {
+			return NULL;
+	}
+	conn->resource_count = resource_count;
+	conn->display = wl_display_connect(NULL);
+	if (!conn->display) {
+		goto fail;
+	}
+	conn->registry = wl_display_get_registry(conn->display);
+	wl_registry_add_listener(conn->registry, &reg_listener, conn);
+	wl_display_roundtrip(conn->display);
+	wl_display_roundtrip(conn->display);
+
+	switch (mode) {
+	case CONSUME_MAPS: {
+		if (!conn->shm) {
+			goto fail;
+		}
+		conn->pools = calloc(resource_count, sizeof(struct wl_shm_pool *));
+		if (!conn->pools) {
+			goto fail;
+		}
+		for (uint32_t i = 0; i < conn->resource_count; i++) {
+			conn->pools[i] = wl_shm_create_pool(conn->shm, fd, POOL_SIZE);
+			/* the roundtrip checks if this connection broke */
+			if (wl_display_roundtrip(conn->display) == -1) {
+				goto fail;
+			}
+		}
+	} break;
+	case CONSUME_FDS: {
+		if (!conn->linux_dmabuf) {
+			goto fail;
+		}
+		conn->buffers = calloc(resource_count, sizeof(struct wl_buffer *));
+		if (!conn->buffers) {
+			goto fail;
+		}
+		for (uint32_t i = 0; i < conn->resource_count; i++) {
+			struct zwp_linux_buffer_params_v1 *params =
+				zwp_linux_dmabuf_v1_create_params(
+					conn->linux_dmabuf);
+			zwp_linux_buffer_params_v1_add(params, fd, 0, 0,
+				DMABUF_WIDTH * 4, 0, 0);
+			conn->buffers[i] = zwp_linux_buffer_params_v1_create_immed(
+				params, DMABUF_WIDTH, DMABUF_HEIGHT, DMABUF_FORMAT, 0);
+			zwp_linux_buffer_params_v1_destroy(params);
+
+			/* the roundtrip checks if this connection broke */
+			if (wl_display_roundtrip(conn->display) == -1) {
+				goto fail;
+			}
+		}
+	} break;
+	case CONSUME_MEMORY: {
+		if (!conn->compositor) {
+			goto fail;
+		}
+		conn->regions = calloc(resource_count / POINTS_PER_REGION + 1,
+			sizeof(struct wl_region *));
+		if (!conn->regions) {
+			goto fail;
+		}
+
+		/* Because most compositors incur quadratic runtime when
+		 * creating regions with many points, divide the points
+		 * registered between regions */
+		struct wl_region *current_region = NULL;
+		uint32_t region_no = 0;
+		uint32_t isqrt = 0;
+		for (uint32_t i = 0; i < conn->resource_count; i++) {
+			if (i >= region_no * POINTS_PER_REGION) {
+				current_region = wl_compositor_create_region(conn->compositor);
+				conn->regions[region_no++] = current_region;
+			}
+			/* add pixels in 50% density, pseudorandom pattern */
+			uint64_t pos_no = 2*i + ((uint64_t)rand()%2);
+			if (pos_no >= (isqrt+1)*(isqrt+1)) {
+				isqrt++;
+			}
+			int32_t x = (int32_t)isqrt, y = (int32_t)(pos_no - isqrt*isqrt);
+			wl_region_add(current_region, x, y, 1, 1);
+
+			if (i % 64 == 0) {
```

Comment:
Is this used to ensure the requests get sent? I wonder if `wl_display_flush` would allow us to reach the limit faster?

Or maybe this is used to make the limit measurement more precise?

Reply:
The periodic `wl_display_roundtrip` calls are for rate limiting, so that `resource-thief` doesn't overflow the buffers of its connection with the compositor, and be disconnected. I've changed this to `i % 128 = 0`, which ensures that the client sends no more than 24 bytes/message * 128 messages = 3072 bytes before the compositor is forced to process them all.

Reply:
Hm, I guess `wl_display_flush` may overflow the kernel internal buffers plus the compositor's libwayland receive buffers, while a roundtrip ensures we don't fill these. I was thinking about libwayland's send buffer only.

Ah, but it doesn't seem like this can happen: `wl_display_flush` is documented to return `errno = EAGAIN` if the remote receive buffers are full. So in theory we could `wl_display_flush`, and if it returns `EAGAIN` wait for the FD to become writable and try again. That would maybe be faster than a roundtrip. But that's probably not worth it if a roundtrip works fine.